File: src/main/java/dev/adamko/gildedrose/java/item/ItemTemplate.java
package dev.adamko.gildedrose.java.item;

import dev.adamko.gildedrose.java.mutators.ItemMutator;
import java.util.Arrays;
import kotlin.collections.CollectionsKt;


/**
 * A Java implementation of {@link dev.adamko.gildedrose.item.ItemTemplate}
 *
 * @see dev.adamko.gildedrose.item.ItemTemplate
 */
public class ItemTemplate implements ItemMutator {

  private final ItemMutator mutator;

  public ItemTemplate(ItemMutator... mutators) {
    assert mutators.length > 0 : "At least one mutator is required, received none.";
    this.mutator = CollectionsKt.reduce(Arrays.asList(mutators), ItemMutator::plus);
  }

  @Override
  public ItemState applyMutation(ItemState item) {
    return mutator.applyMutation(item);
  }
}

File: src/main/kotlin/dev/adamko/gildedrose/item/ItemTemplate.kt
package dev.adamko.gildedrose.item


/**
 * Defines how a specific instance of an [ItemState] should be handled.
 *
 * Holds functions for the mutation of [ItemState]s.
 *
 * A template can be associated with a template with [ItemTemplateProvider].
 *
 * See [ItemTemplates] for implementations.
 *
 * @see ItemTemplates
 */
open class ItemTemplate(
    private val mutator: ItemMutator,
) : ItemMutator by mutator {

  constructor(
      vararg mutators: ItemMutator
  ) : this(
      if (mutators.isEmpty())
        throw IllegalStateException("At least one mutator is required, received none.")
      else
        mutators.asSequence().reduce(ItemMutator::plus)
  )

}
